Add HIV test results

mpbrown opened the pull request:
# FRONTEND PULL REQUEST

## Related Issue

- #6082

## Changes Proposed

- Test results shown on the test card will now be data driven so that all supported disease tests for the selected device are shown
- Individual diseases like HIV or RSV will be filtered out if their feature flags are not enabled
- Validation requirements for test results have been simplified so that the user can choose which diseases they report a result for or leave blank, as long as **at least one result is selected**
- This means multiplex for COVID, Flu A, and Flu B
    - will no longer require the user to report both flu if only one of the flu results is reported
    - will allow the user to select mark individual results as inconclusive without having to mark all results as inconclusive

## Additional Information

- Verified with product that we don't actually need the complex validation requirements for individual disease combinations like multiplex devices
- Will be following this up with another PR that adds HIV AOE

## Testing

- Verify that the test card still works as expected 
- Deployed on [dev2](https://dev2.simplereport.gov/)

## Screenshots / Demos

### HIV only device
![Screenshot 2023-11-03 104754](https://github.com/CDCgov/prime-simplereport/assets/23287037/0ec36857-d470-4042-b580-c65a76d6902e)

### Covid, Flu A, Flu B multiplex
![Screenshot 2023-11-03 104808](https://github.com/CDCgov/prime-simplereport/assets/23287037/ed63120e-9c34-4ab9-b961-1820b29b5727)

### Covid only device
![Screenshot 2023-11-03 104818](https://github.com/CDCgov/prime-simplereport/assets/23287037/f61a3474-bc4f-48b4-80b3-66fae9c77014)

<!---
## Checklist for Author and Reviewer
### Accessibility
- [ ] Any large changes have been run through Deque manual testing
- [ ] All changes have run through the Deque automated testing

### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### Content
- [ ] Any content changes have been approved by content team

### Support
- [ ] Any changes that might generate new support requests have been flagged to the support team
- [ ] Any changes to support infrastructure have been demo'd to support team

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

### Documentation
- [ ] Any changes to the startup configuration have been documented in the README
-->

mpbrown commented on frontend/src/app/testQueue/TestCard/TestCard.test.tsx:
```diff
@@ -262,9 +271,42 @@ describe("TestCard", () => {
+    deviceNameOrder: string[] = DEFAULT_DEVICE_ORDER
+  ) => {
+    deviceNameOrder.forEach((deviceName, index) => {
+      expect(deviceDropdown.options[index].label).toEqual(deviceName);
```
Many tests were checking the order of the device options and the individual array indices would need to be updated any time there is a new device added to the tests so I just extracted this to its own function

mpbrown commented on frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/TestResultInputGroup.tsx:
```diff
@@ -0,0 +1,123 @@
+    return (
+      <Alert headingLevel={"h3"} type={"error"}>
+        This device doesn't have any supported disease tests configured. Please
+        ask your organization admin to add the correct test.
```
Checking with design and content to see how we want to handle this error case where the device doesn't have any diseases configured, but this seems to be an edge case only encountered on some of the dev environments that still have pre-multiplex devices in the database

fzhao99 commented on frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:
```diff
@@ -249,17 +239,32 @@ const TestCardForm = ({
+      const isFilled =
+        result.testResult === TEST_RESULTS.POSITIVE ||
+        result.testResult === TEST_RESULTS.NEGATIVE ||
+        result.testResult === TEST_RESULTS.UNDETERMINED;
```
could we just use `result.testResult !== TEST_RESULTS.UNKNOWN` here? [Reading the enum definition ](https://github.com/CDCgov/prime-simplereport/blob/main/frontend/src/app/testResults/constants.ts) those two seem to be equivalent

mpbrown replied:
The `MultiplexResultInput` type allows test result to be undefined, so I'd lean toward keeping these to be safe. Also realized from this that the test results aren't being explicitly set as `TEST_RESULTS.UNKNOWN` in the form state so I'll update that

fzhao99 commented on frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:
```diff
@@ -249,17 +239,32 @@ const TestCardForm = ({
   };
 
   const validateTestResults = () => {
```
Any chance we could rename these and the related `validate....` functions to something like `generateTestResultsErrors`? It felt weird to me that we were returning an empty string in a validation function like this (would have expected a boolean / undefined / null return in the case of this is / isn't valid) but see instead that we're using these to generate an error state, which makes me think that these functions are misnamed. 

Along the same lines, I think it might be more readable to return something like `null` here and then do an explicit null check on 271 and related places. That would divorce the content of the error from the state of whether the error exists or not which aren't the same. The check would read something like
```
const testResultErrors = generateTestResultErrors()
const showTestResultsError = hasAttemptedSubmit && testResultsError != null;
....
   if (testResultsError) { // or if(testResultsError !== null)
      showError(testResultsError, "Invalid test results");
    }
```

mpbrown replied:
Yeah I went back and forth on whether the name should mention validation. Does `getTestResultsError` and `getDateTestedError` sound good? I'll also update those functions to return `null` instead of empty string

fzhao99 commented on frontend/src/app/testQueue/TestCardForm/TestCardForm.tsx:
```diff
@@ -323,7 +328,8 @@ const TestCardForm = ({
         specimenTypeId: state.specimenId,
         dateTested: state.dateTested,
-        results: doesDeviceSupportMultiplex(state.deviceId, devicesMap)
+        // Should we filter results to only send results supported by the selected device?
```
could you elaborate a bit more about how this described state differs from what we're already doing? In the case where `doesDeviceSupportNonCovid` and we're not already doing filtering, presumably `state.testResults` only has defined values for the diseases the devices support. Are we returning nulls / undefineds for the other diseases that we'd want to potentially filter out?

mpbrown replied:
I think my question in that comment might actually be addressed by clearing the state of test results whenever a device changes. On the old test card, if you marked Flu A as positive and then changed the device to a COVID only device, it would technically still have the Flu A value in the test results state which would just get filtered out while submitting

fzhao99 commented on frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/TestResultInputGroup.tsx:
```diff
@@ -0,0 +1,123 @@
+type TestResultRecord = Record<string, TEST_RESULTS>;
+
+const convertTestResultRecordToArray = (testResultRecord: TestResultRecord) => {
+  return Object.entries(testResultRecord).map((entry): MultiplexResultInput => {
```
could we just directly spread out entry into its parts using 

```
// could also do something like .map(([key, value]))
.map(([diseaseName, testResult]): MultiplexResultInput => {
    return {
      diseaseName: diseaseName,
      testResult: testResult,
    };
  });
``` here?

fzhao99 commented:
Did a first pass and left some questions! Good work on this

fzhao99 reviewed (changes requested):
Did a first pass and left some questions! One issue in particular for cases where devices may have multiple configurations of the same disease that needs to be handled 

Good work on this!

mpbrown commented on frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/TestResultInputGroup.tsx:
```diff
@@ -0,0 +1,123 @@
+  devicesMap,
+  onChange,
+}: Props) => {
+  const supportedDiseaseTests =
```
Thanks for finding this case! We might be able to use `test performed code` to determine whether the diseases are identical and filter it out of the list, but I'm checking with product on whether that's a reliable field to use.

I don't think it'll necessarily matter which COVID we use in this case since the test card just sends the disease name and test result along with the selected device instead of sending anything specific to the `test performed code` or `test ordered code`

mpbrown replied:
Just pushed a commit that should filter this out based on the duplicate disease name

emyl3 commented on frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/TestResultInputGroup.tsx:
```diff
@@ -0,0 +1,145 @@
+  onChange,
+}: Props) => {
+  const unfilteredDiseaseTests =
+    devicesMap.get(deviceId)?.supportedDiseaseTestPerformed ?? [];
```
we aren't filtering out rsv and hiv based on the feature flag so they are showing up as result options in the test card (like how we are filtering them out in the device options)

mpbrown replied:
Great find on this! I just pushed a fix and will redeploy on dev2

emyl3 reviewed (changes requested):
Left one thing to change/investigate! Nice work, Mike!! 😸

mpbrown commented:
Updated to filter out RSV and HIV from the test result options if those feature flags are turned off. These screenshots are from my local deployment with the `singleEntryRsvEnabled` flag set to false

### Multiplex device with RSV as a supported disease
![Screenshot 2023-11-07 150704](https://github.com/CDCgov/prime-simplereport/assets/23287037/d0d97a5b-1fcc-4592-b8aa-aae068ac0185)

### Test card using that device but filtering out RSV
![Screenshot 2023-11-07 150718](https://github.com/CDCgov/prime-simplereport/assets/23287037/8040e484-ce14-425c-bca5-17e31e2927c4)

emyl3 reviewed (commented):
The filtering on the test card looks good!

If this change enables the new test card for all users, I would like to make sure the pregnancy bug that @mehansen discovered is fixed before it goes in.

Since I think that is affecting the ability to keyboard tab to the pregancy AOE question.
To repro:
- have a test card in the queue where the pregnancy AOE question was not answered. 
- try "tabbing" to the pregnancy question
- focus will skip over the question

I also noticed one accessibility concern when running the axe dev tool
<img width="921" alt="Screenshot 2023-11-07 at 14 35 29" src="https://github.com/CDCgov/prime-simplereport/assets/20211771/f079ca75-06bd-41f2-b602-3b2623091b99">

I'm also unable to run a couple of the guided tests on this page (might be a bug on their end -- need to investigate 🧐)

Thank you for all your work on this Mike!

changecourse commented:
Things are generally looking good from a styling perspective, but I did notice some things (when comparing to the [figma instance](https://www.figma.com/file/LglkDsGZDKMZm3XCkJPhfR/SimpleReport-2023%2B?type=design&node-id=3285%3A12803&mode=design&t=wMiXgEGfvpGQ3WBR-1))

- The order of the symptoms question vs. the pregnancy question is swapped compared to design (may not be incorrect, just calling it out)

- The spacing between questions should be a consistent 24px (seems shorter between questions 3 and 4

- The radio button labels for COVI-19 on the "Test card using that device but filtering out RSV" appear to be smaller than all other radio button labels

- The padding between columns of radio button groups on multiplex card use cases should be greater for readability; 60px between columns is sufficient

mpbrown commented:
@emyl3 I refactored the test queue to use a list element, but I can't reproduce the pregnancy bug so it might be best for us to fix that in a different ticket


https://github.com/CDCgov/prime-simplereport/assets/23287037/96606048-ce83-42e6-ad45-fd8bbd61b5ab


@changecourse regarding the order of pregnancy and symptoms, I used Kenny's [Figma file here](https://www.figma.com/file/LglkDsGZDKMZm3XCkJPhfR/SimpleReport-2023%2B?type=design&node-id=1707-7291&mode=design&t=SWUyiXI93d7eCGfG-0) which has symptoms at the end because it expands with the additional questions if the user selected Yes.

I updated the spacing between the questions and between the columns of radio button groups. The radio button labels have the same height dimensions and font sizes

![Screenshot 2023-11-07 172343](https://github.com/CDCgov/prime-simplereport/assets/23287037/9cf97c3d-6522-448d-8b3c-0a5a3f82ea13)
![Screenshot 2023-11-07 172336](https://github.com/CDCgov/prime-simplereport/assets/23287037/9c9793ac-b578-48b4-a3c7-686a8cee71a4)

emyl3 commented:
@mpbrown Sorry should have left steps to repro the pregnancy bug!
You need two tests in the queue.
- one test leaves the pregnancy question blank
- other test selects an answer
- tabbing to the pregnancy question for the test that left it blank does not work

After doing some digging...
I think this is happening because the name for the radio group is not unique:
https://github.com/CDCgov/prime-simplereport/blob/ff837fa096f49fbe0baf5d504ec4629295d195bb/frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/CovidAoEForm.tsx#L104

compare that with the has symptoms question that works:
https://github.com/CDCgov/prime-simplereport/blob/ff837fa096f49fbe0baf5d504ec4629295d195bb/frontend/src/app/testQueue/TestCardForm/diseaseSpecificComponents/CovidAoEForm.tsx#L114

Making it unique should fix the issue!

emyl3 reviewed (approved):
LGTM! Thank you for all your work on this, Mike!

I created a separate bug ticket (I thought this was resolved before but maybe it was for something very similar 🤔 😅 ) for an edge case I noticed on dev2 while testing: #6940

fzhao99 reviewed (approved):
lgtm! thank you for the effort
